### PR TITLE
Migrate test suite to GoogleTest framework

### PR DESCRIPTION
**Pull request by davidtrevelyan**
This PR does the following:

- Introduces GoogleTest as the testing framework for RTNeural
- Introduces two test targets: `rtneural_test_unit` and `rtneural_test_functional`
  - Unit tests contain a small new example demonstrating unit testing patterns for `Activation` layers
  - Functional tests are simply the old test header files, moved to the new target. Some old assertions have been simplified to make use of test assertions built into GoogleMock, and there have been a couple of very minor clarifying refactors in the test code. No new functional tests have been added. 
- Adds a new README file in the `tests` folder, explaining how to build and run the tests (or a subset of them!)
- Updates test implementations to use `#include <RTNeural/RTNeural.h>`, to match the intended client code inclusion pattern

### CODE REVIEW

**Review by jatinchowdhury18 — APPROVED** [review]
Thanks for the PR! The changes look good overall, I just had a couple of notes.
- For the RTNeural header, I think using `#include <RTNeural/RTNeural.h>` in both the tests and client code is the way to go.
- I think there's some kind of linker error happening with `gmock` on MSVC. I think this can be fixed by adding this to our test targets:
```cmake
set_property(TARGET foo PROPERTY
  MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```

File: cmake/Testing.cmake [hunk 1]
```diff
@@ -0,0 +1,31 @@
+option(RTNEURAL_TEST_REPORTS "Output test reports to XML files" OFF)
+
+macro(rtneural_setup_testing)
+    include(CTest)
```
[the remark this answers is not in the document]
Comment (davidtrevelyan) [reply cmake/Testing.cmake:4]: I believe you're correct on both counts :)

A `macro` in CMake is somewhat analogous to a C preprocessor macro - after the arguments are substituted into the body, it's as if the implementation of the macro is just copy/pasted into the call site. This particular macro has no arguments, so it's pretty much just a straight copy/paste of the macro body! The result is as you suspect - CTest is indeed included at the top level CMake script here 👍

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# RTNeural Testing
+
+RTNeural tests are configured with CMake's [CTest](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html) and developed with the [GoogleTest framework](https://github.com/google/googletest). Tests are split into two categories:
+
+1. unit tests, and
+2. functional tests.
```
[the remark this answers is not in the document]
Comment (davidtrevelyan) [reply tests/README.md:6]: Glad you like it! On splitting the functional tests into tests for layers vs models, I'm not sure. I'd be happy to jump on a call to chat about it with you at some point, if you like! I would consider whether the layer tests are simple enough to be written as unit tests - they're usually simpler, easier to reason about, and less brittle. But every project has different needs and challenges, so there's no one-size-fits-all :)

File: RTNeural/CMakeLists.txt [hunk 3]
```diff
@@ -44,7 +44,7 @@ target_include_directories(RTNeural
     PUBLIC
         ../modules/json
     INTERFACE
-        ..
+        ./
```
[the remark this answers is not in the document]
Comment (davidtrevelyan) [reply RTNeural/CMakeLists.txt:47]: For sure - sounds great. Totally agree on avoiding a breaking change. I'll update the tests and revert the include path!

[the remark this answers is not in the document]
Comment (davidtrevelyan) [reply RTNeural/CMakeLists.txt]: This one's done, and I've updated the PR description to match!

**Comment by davidtrevelyan** [comment]
> Thanks for the PR! The changes look good overall, I just had a couple of notes.
> 
> * For the RTNeural header, I think using `#include <RTNeural/RTNeural.h>` in both the tests and client code is the way to go.
> * I think there's some kind of linker error happening with `gmock` on MSVC. I think this can be fixed by adding this to our test targets:
> 
> ```cmake
> set_property(TARGET foo PROPERTY
>   MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
> ```

No problem - thanks for the review! 

- Regarding the include directories and include style, I've reverted the target property and have updated the tests to use the new style, and
- regarding the gmock linker error - oops! Missed that one. I think the solution to this is in the [GoogleTest README](https://github.com/google/googletest/blob/main/googletest/README.md#visual-studio-dynamic-vs-static-runtimes), so I'll try that first. If it doesn't work, let's try doing it manually as you suggest?

**Comment by jatinchowdhury18** [comment]
Awesome, thanks for the update! Yeah, trying the CMake option provided by GoogleTest sounds good.

I think there's also a different include path thing happening in some of the examples, but otherwise we should be good to go :)

**Comment by davidtrevelyan** [comment]
Nice one, thanks for the excellent review @jatinchowdhury18! I've updated the contributors list now and I'm ready for it to be merged too, assuming the CI passes 😁

**Pull request merged**